modules/tinymce/src/themes/silver/test/ts/webdriver/components/urlinput/UrlInputTest.ts

```diff
@@ -0,0 +1,78 @@
+import { beforeEach, context, describe, it } from '@ephox/bedrock-client';
+import { Future, Optional } from '@ephox/katamari';
+import { SelectorFind, SugarDocument, Value } from '@ephox/sugar';
+import { assert } from 'chai';
+
+import { ApiUrlData } from 'tinymce/themes/silver/backstage/UrlInputBackstage';
+import { renderUrlInput } from 'tinymce/themes/silver/ui/dialog/UrlInput';
+
+import * as TestExtras from '../../../module/TestExtras';
+
+describe('webdriver.tinymce.themes.silver.components.urlinput.UrlInputTest', () => {
+  const helpers = TestExtras.bddSetup();
+
+  const hook = TestHelpers.GuiSetup.bddSetup((store, _doc, _body) => GuiFactory.build(
+    renderUrlInput({
+      label: Optional.some('UrlInput label'),
+      name: 'col1',
+      filetype: 'file',
+      enabled: true
+    }, helpers.backstage(), {
+      getHistory: (_fileType) => [],
+      addToHistory: (_url, _filetype) => store.adder('addToHistory')(),
+      getLinkInformation: () => Optional.none(),
+      getValidationHandler: () => Optional.none(),
+      getUrlPicker: (_filetype) => Optional.some((entry: ApiUrlData) => {
+        store.adder('urlpicker')();
+        return Future.pure({ value: 'http://tiny.cloud', meta: { before: entry.value }, fieldname: 'test' });
+      })
+    }, Optional.none())
+  ));
+
+  TestHelpers.GuiSetup.bddAddStyles(SugarDocument.getDocument(), [
+    '.tox-menu { background: white; }',
+    '.tox-collection__item--active { background: #cadbee }'
+  ]);
+
+  const getInput = () => {
+    const component = hook.component();
+    return component.getSystem().getByDom(
+      SelectorFind.descendant(component.element, 'input').getOrDie('Could not find input')
+    ).getOrDie();
+  };
+
+  const assertTrimSpaces = (inputKeys: any[], expectedOutput: string) =>
+    async () => {
+      const input = getInput();
+
+      await RealKeys.pSendKeysOn('body input', inputKeys);
+
+      assert.equal(Value.get(input.element), expectedOutput, 'Checking Value.get');
+      const repValue = Representing.getValue(input);
+      assert.deepEqual(
+        {
+          value: repValue.value,
+          meta: { text: repValue.meta.text },
+        },
+        {
+          value: expectedOutput,
+          meta: { text: undefined },
+        },
+        'Checking Rep.getValue'
+      );
+    };
+
+  beforeEach(() => {
+    hook.store().clear();
+    UiControls.setValue(getInput().element, '');
+  });
+
+  context('Trim spaces as expected for an URL', () => {
+    it('TINY-8775: Trim a space of it is the only thing in the input', assertTrimSpaces([ RealKeys.text(' ') ], ''));
+    it('TINY-8775: Trim a space of it is at the start', assertTrimSpaces([ RealKeys.text(' A') ], 'A'));
+    it('TINY-8775: Trim a space of it is at the end', assertTrimSpaces([ RealKeys.text('A ') ], 'A'));
+    it('TINY-8775: Trim all appropriate spaces', assertTrimSpaces([ RealKeys.text(' A B '), RealKeys.combo({}, 'arrowLeft'), RealKeys.text(' ') ], 'A B'));
```

Comment:
Does this really result in `A B` when typing one character at a time on a keyboard? I'm not sure the input filter logic would really allow for that if it trims `"A "` before the `B` is typed. I assume `RealKeys.text()` is just inserting all at once as if it was pasted.

Reply:
Bedrock uses `elementSendKeys` which will handle them one character at a time as per https://w3c.github.io/webdriver/#dfn-dispatch-the-events-for-a-typeable-string. So typing ` A B ` results in `AB` being in the output instead of it being similar to being pasted and that can be verified by removing the other keys here. So after the state is `AB`, pressing left then moves the caret between `A` and `B` and finally pressing space adds the space in the middle so it ends up as `A B`.

Reply:
ok. thanks for the update.